AnnoDesigner/AnnoCanvas.xaml.cs

```diff
@@ -29,6 +29,144 @@
+        void UndoAction(List<LayoutObject> placedObjects);
+    }
+
+    /// <summary>
+    /// Abstract base class that implements adding or removing objects on the canvas.
+    /// </summary>
+    public abstract class AddRemovePlacedObjects : IAction
+    {
+        private List<LayoutObject> _objectsToPlace;
+        private AddRemove _addOrRemove;
+
+        protected enum AddRemove {
+            Add,
+            Remove
+        }
+
+        protected AddRemovePlacedObjects(List<LayoutObject> objectsToPlace, AddRemove addOrRemove)
+        {
+            _objectsToPlace = objectsToPlace;
+            _addOrRemove = addOrRemove;
+        }
+
+        public void PerformAction(List<LayoutObject> placedObjects)
+        {
+            switch (_addOrRemove)
+            {
+                case AddRemove.Add:
+                    addObject(placedObjects);
+                    break;
+                case AddRemove.Remove:
+                    removeObject(placedObjects);
+                    break;
+            }
+        }
+
+        public void UndoAction(List<LayoutObject> placedObjects)
+        {
+            switch (_addOrRemove) 
+            {
+                case AddRemove.Add:
+                    removeObject(placedObjects);
+                    break;
+                case AddRemove.Remove:
+                    addObject(placedObjects);
+                    break;
+            }
+
+        }
+
+        private void addObject(List<LayoutObject> placedObjects)
+        {
+            placedObjects.AddRange(_objectsToPlace);
+            // sort the objects because borderless objects should be drawn first
+            placedObjects.Sort((a, b) => b.WrappedAnnoObject.Borderless.CompareTo(a.WrappedAnnoObject.Borderless));
+        }
+
+        private void removeObject(List<LayoutObject> placedObjects)
+        {
+            // placedObjects = placedObjects.Except(_objectsToPlace).ToList();
+            _objectsToPlace.ForEach(_ => placedObjects.Remove(_));
+        }
+    }
+
+    /// <summary>
+    /// Places objects on the canvas.
+    /// </summary>
+    public class AddPlacedObjects : AddRemovePlacedObjects
+    {
+        public AddPlacedObjects(List<LayoutObject> objectsToPlace) : base(objectsToPlace, AddRemove.Add) { }
+    }
+
+    /// <summary>
+    /// Removes objects from the canvas.
+    /// </summary>
+    public class RemovePlacedObjects : AddRemovePlacedObjects
+    {
+        public RemovePlacedObjects(List<LayoutObject> objectsToPlace) : base(objectsToPlace, AddRemove.Remove) { }
+    }
+
+    /// <summary>
+    /// Class which manages the actions that have been taken. Allows performing new actions, and undoing and redoing actions.
+    /// </summary>
+    public class ActionManager
+    {
+        private static readonly Logger logger = LogManager.GetCurrentClassLogger();
+
+        private List<LayoutObject> _placedObjects;
+        private List<IAction> _previousActions;
```

Comment:
Use stacks instead.

Reply:
Good idea, not familiar with available data structures. Popping off the list felt really cumbersome.